src/rng/src/main.rs

```diff
@@ -0,0 +1,192 @@
+// VIRTIO RNG Emulation via vhost-user
+//
+// Copyright 2021 Linaro Ltd. All Rights Reserved.
+// Mathieu Poirier <mathieu.poirier@linaro.org>
+//
+// SPDX-License-Identifier: Apache-2.0
+
+mod vhu_rng;
+
+use clap::{load_yaml, App};
+use std::fs::File;
+use std::path::Path;
+use std::sync::{Arc, Mutex, RwLock};
+use std::thread;
+use vhost::{vhost_user, vhost_user::Listener};
+use vhost_user_backend::VhostUserDaemon;
+use vhu_rng::VuRngBackend;
+use vm_memory::{GuestMemoryAtomic, GuestMemoryMmap};
+
+fn start_backend(cmd_args: clap::ArgMatches, dry_run: bool) -> Result<(), String> {
+    let mut handles = Vec::new();
+
+    let path = cmd_args
+        .value_of("socket_path")
+        .ok_or("Invalid socket path")?;
+
+    let count = cmd_args
+        .value_of("socket_count")
+        .unwrap_or("1")
+        .parse::<u32>()
+        .map_err(|_| "Invalid socket_count")?;
```

Comment:
Hi @andreeaflorescu,

Can you point me to a specific example?   That way there is no misunderstanding about what you are looking for.

Reply:
You'll need to define an enum as follows, and then change the signature of the `start_backend` function to return a `Result<(), BackendError>` instead of errors as strings.

This approach will later help with writing tests for the error cases as well because we will be able to compare sending invalid input to the function with an error variant, instead of using `is_err`.

```rust
fn start_backend(cmd_args: clap::ArgMatches, dry_run: bool) -> Result<(), BackendError> {
    let mut handles = Vec::new();

    let path = cmd_args
        .value_of("socket_path")
        .ok_or(BackendError::InvalidSocketPath)?;

    let count = cmd_args
        .value_of("socket_count")
        .unwrap_or("1")
        .parse::<u32>()
        .map_err(|_| BackendError::InvalidSocketCount)?;
.......
```

Then, in unit tests, you can write:

```rust
    #[test]
    fn invalid_random_file_input() {
        assert_eq!(start_backend(get_cmd_args("-f", "/dev/doesnotexists"), true), BackendError::InvalidSocketPath);
    }
```

You'll want to also implement the Display & Error traits. You can find some more examples here: https://doc.rust-lang.org/std/error/trait.Error.html